bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCallbackUtils.java

```diff
@@ -0,0 +1,250 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.bookkeeper.client;
+
+import com.google.common.collect.Iterators;
+import java.util.Enumeration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import static org.apache.bookkeeper.client.LedgerHandle.LOG;
+
+/**
+ * Utility for callbacks
+ *
+ */
+class SyncCallbackUtils {
+
+    /**
+     * Wait for a result. This is convenience method to implement callbacks
+     *
+     * @param <T>
+     * @param future
+     * @return
+     * @throws InterruptedException
+     * @throws BKException
+     */
+    public static <T> T waitForResult(CompletableFuture<T> future) throws InterruptedException, BKException {
+        try {
+            return future.get();
+        } catch (ExecutionException err) {
+            if (err.getCause() instanceof BKException) {
+                throw (BKException) err.getCause();
+            } else {
+                BKException unexpectedConditionException
+                    = BKException.create(BKException.Code.UnexpectedConditionException);
+                unexpectedConditionException.initCause(err.getCause());
+                throw unexpectedConditionException;
+            }
+
+        }
+    }
+
+    /**
+     * Handle the Response Code and transform it to a BKException
+     *
+     * @param <T>
+     * @param rc
+     * @param result
+     * @param future
+     */
+    public static <T> void finish(int rc, T result, CompletableFuture<T> future) {
+        if (rc != BKException.Code.OK) {
+            future.completeExceptionally(BKException.create(rc).fillInStackTrace());
+        } else {
+            future.complete(result);
+        }
+    }
+
+    static class SyncCreateCallback extends CompletableFuture
```

Comment:
@sijie it is just a problem of types
SyncCreateCallback is used to create LedgerHandle, LedgerHandleAdv, WriteHandle and WriteAdvHandle
the hierarchy is a bit complex.
A cleaner approch would be to have 4 different classes, maybe it is the best, it does not introduce costs at runtime and it is really simple

Reply:
why can't you use `super`? I'd prefer enforcing type safe.


```
static class SyncCreateCallback {

    private final CompletableFuture<? super LedgerHandle> future;

   SyncCreateCallback(CompletableFuture<? super LedgerHandle> future) {
       this.future = future;
   }
   
    public void createComplete(int rc, LedgerHandle lh, Object ctx) {
         SyncCallbackUtils.finish(rc, lh, future);
    }

}
```

so you can use it for different interfaces.

```
Completable<LedgerHandle> future = new CompletableFuture();
SyncCreateCallback callback = new SyncCreateCallback(future);

Completable<WriteHandle> future = new CompletableFuture<>();
SyncCreateCallback callback = new SyncCreateCallback(future);
```

Reply:
@sijie I will take your approach, I did not see your comments and I had already committed a proposal.
I will be back on this work soon
thank you

Reply:
>> CompletableFuture<? super LedgerHandle> future

this does not work because WriteHandle is not a subclass of LedgerHandle
I am updating the patch with a new tradeoff